File: src/services/notification.service.ts
import { apiClient } from '@/lib/api-client';
import { 
  Notification, 
  NotificationResponse, 
  NotificationStats, 
  NotificationPreferences,
  CreateNotificationRequest,
  MarkNotificationRequest,
  NotificationFilters
} from '@/types/notification';

export class NotificationService {
  private static readonly BASE_URL = '/notifications';

  // Lấy danh sách thông báo của user hiện tại
  static async getNotifications(filters?: NotificationFilters): Promise<NotificationResponse> {
    const params = new URLSearchParams();
    
    if (filters?.isRead !== undefined) {
      params.append('isRead', filters.isRead.toString());
    }
    if (filters?.type) {
      params.append('type', filters.type);
    }
    if (filters?.category) {
      params.append('category', filters.category);
    }
    if (filters?.dateFrom) {
      params.append('dateFrom', filters.dateFrom);
    }
    if (filters?.dateTo) {
      params.append('dateTo', filters.dateTo);
    }
    if (filters?.page) {
      params.append('page', filters.page.toString());
    }
    if (filters?.size) {
      params.append('size', filters.size.toString());
    }

    const queryString = params.toString();
    const url = queryString ? `${this.BASE_URL}?${queryString}` : this.BASE_URL;
    
    return apiClient.get<NotificationResponse>(url);
  }

  // Lấy thống kê thông báo
  static async getNotificationStats(): Promise<NotificationStats> {
    return apiClient.get<NotificationStats>(`${this.BASE_URL}/stats`);
  }

  // Lấy số lượng thông báo chưa đọc
  static async getUnreadCount(): Promise<{ count: number }> {
    return apiClient.get<{ count: number }>(`${this.BASE_URL}/unread-count`);
  }

  // Đánh dấu thông báo đã đọc/chưa đọc
  static async markNotifications(request: MarkNotificationRequest): Promise<void> {
    return apiClient.put<void>(`${this.BASE_URL}/mark`, request);
  }

  // Đánh dấu tất cả thông báo là đã đọc
  static async markAllAsRead(): Promise<void> {
    return apiClient.put<void>(`${this.BASE_URL}/mark-all-read`);
  }

  // Xóa thông báo
  static async deleteNotification(notificationId: number): Promise<void> {
    return apiClient.delete<void>(`${this.BASE_URL}/${notificationId}`);
  }

  // Xóa nhiều thông báo
  static async deleteNotifications(notificationIds: number[]): Promise<void> {
    return apiClient.request<void>(`${this.BASE_URL}/bulk`, {
      method: 'DELETE',
      body: JSON.stringify({ notificationIds }),
    });
  }

  // Lấy cài đặt thông báo
  static async getNotificationPreferences(): Promise<NotificationPreferences> {
    return apiClient.get<NotificationPreferences>(`${this.BASE_URL}/preferences`);
  }

  // Cập nhật cài đặt thông báo
  static async updateNotificationPreferences(
    preferences: Partial<NotificationPreferences>
  ): Promise<NotificationPreferences> {
    return apiClient.put<NotificationPreferences>(`${this.BASE_URL}/preferences`, preferences);
  }

  // Tạo thông báo (dành cho admin/teacher)
  static async createNotification(request: CreateNotificationRequest): Promise<Notification> {
    return apiClient.post<Notification>(`${this.BASE_URL}`, request);
  }

  // Gửi thông báo đến tất cả sinh viên trong khóa học
  static async sendNotificationToCourse(
    courseId: number, 
    request: CreateNotificationRequest
  ): Promise<void> {
    return apiClient.post<void>(`${this.BASE_URL}/course/${courseId}`, request);
  }

  // Gửi thông báo theo role
  static async sendNotificationByRole(
    role: 'STUDENT' | 'TEACHER' | 'ADMIN',
    request: CreateNotificationRequest
  ): Promise<void> {
    return apiClient.post<void>(`${this.BASE_URL}/role/${role}`, request);
  }

  // Lấy thông báo theo ID
  static async getNotificationById(notificationId: number): Promise<Notification> {
    return apiClient.get<Notification>(`${this.BASE_URL}/${notificationId}`);
  }

  // Subscribe to real-time notifications (WebSocket)
  static subscribeToNotifications(
    userId: number, 
    onNotification: (notification: Notification) => void
  ): WebSocket | null {
    // TODO: Implement WebSocket when backend WebSocket endpoint is ready
    console.log('WebSocket notifications will be implemented later', { userId });
    return null;
    
    // try {
    //   // Sử dụng WebSocket để nhận thông báo real-time
    //   const wsUrl = `${process.env.NEXT_PUBLIC_WS_URL || 'ws://localhost:8086'}/ws/notifications?userId=${userId}`;
    //   const ws = new WebSocket(wsUrl);

    //   ws.onmessage = (event) => {
    //     try {
    //       const notification: Notification = JSON.parse(event.data);
    //       onNotification(notification);
    //     } catch (error) {
    //       console.error('Error parsing notification:', error);
    //     }
    //   };

    //   ws.onerror = (error) => {
    //     console.error('WebSocket error:', error);
    //   };

    //   return ws;
    // } catch (error) {
    //   console.error('Error setting up WebSocket:', error);
    //   return null;
    // }
  }
}

export const notificationService = NotificationService;

File: cscore-backend/src/main/java/iuh/fit/cscore_be/service/NotificationService.java
package iuh.fit.cscore_be.service;

import iuh.fit.cscore_be.dto.request.CreateNotificationRequest;
import iuh.fit.cscore_be.dto.request.MarkNotificationRequest;
import iuh.fit.cscore_be.dto.response.NotificationPageResponse;
import iuh.fit.cscore_be.dto.response.NotificationResponse;
import iuh.fit.cscore_be.dto.response.NotificationStatsResponse;
import iuh.fit.cscore_be.entity.Assignment;
import iuh.fit.cscore_be.entity.Course;
import iuh.fit.cscore_be.entity.Enrollment;
import iuh.fit.cscore_be.entity.Notification;
import iuh.fit.cscore_be.entity.User;
import iuh.fit.cscore_be.enums.NotificationCategory;
import iuh.fit.cscore_be.enums.NotificationType;
import iuh.fit.cscore_be.enums.Role;
import iuh.fit.cscore_be.repository.AssignmentRepository;
import iuh.fit.cscore_be.repository.CourseRepository;
import iuh.fit.cscore_be.repository.EnrollmentRepository;
import iuh.fit.cscore_be.repository.NotificationRepository;
import iuh.fit.cscore_be.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class NotificationService {

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private CourseRepository courseRepository;

    @Autowired
    private AssignmentRepository assignmentRepository;

    @Autowired
    private EnrollmentRepository enrollmentRepository;

    /**
     * Get notifications for current user with filters
     */
    public NotificationPageResponse getNotifications(
            User currentUser,
            Boolean isRead,
            NotificationType type,
            NotificationCategory category,
            LocalDateTime dateFrom,
            LocalDateTime dateTo,
            int page,
            int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<Notification> notificationPage = notificationRepository.findByUserWithFilters(
                currentUser, isRead, type, category, dateFrom, dateTo, pageable
        );

        // Convert to response DTOs
        Page<NotificationResponse> responsePage = notificationPage.map(NotificationResponse::from);
        
        // Get unread count
        long unreadCount = notificationRepository.countByUserAndIsReadFalse(currentUser);
        
        return NotificationPageResponse.from(responsePage, unreadCount);
    }

    /**
     * Get notification statistics for current user
     */
    public NotificationStatsResponse getNotificationStats(User currentUser) {
        LocalDateTime todayStart = LocalDateTime.now().toLocalDate().atStartOfDay();
        LocalDateTime weekStart = LocalDateTime.now().minusDays(7);
        
        Object[] stats = notificationRepository.getNotificationStats(currentUser, todayStart);
        
        if (stats != null && stats.length >= 3) {
            long totalCount = ((Number) stats[0]).longValue();
            long unreadCount = ((Number) stats[1]).longValue();
            long todayCount = ((Number) stats[2]).longValue();
            
            // Calculate week count
            long weekCount = notificationRepository.countByUserAndDateRange(
                    currentUser, weekStart, LocalDateTime.now()
            );
            
            return new NotificationStatsResponse(totalCount, unreadCount, todayCount, weekCount);
        }
        
        return new NotificationStatsResponse(0, 0, 0, 0);
    }

    /**
     * Get unread notification count
     */
    public long getUnreadCount(User currentUser) {
        return notificationRepository.countByUserAndIsReadFalse(currentUser);
    }

    /**
     * Mark notifications as read/unread
     */
    public void markNotifications(User currentUser, MarkNotificationRequest request) {
        notificationRepository.updateReadStatusByIds(
                request.getNotificationIds(),
                request.getIsRead(),
                currentUser
        );
    }

    /**
     * Mark all notifications as read for current user
     */
    public void markAllAsRead(User currentUser) {
        notificationRepository.markAllAsReadForUser(currentUser);
    }

    /**
     * Delete a notification (only if it belongs to current user)
     */
    public void deleteNotification(User currentUser, Long notificationId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Notification not found"));
        
        if (!notification.getUser().getId().equals(currentUser.getId())) {
            throw new RuntimeException("You don't have permission to delete this notification");
        }
        
        notificationRepository.delete(notification);
    }

    /**
     * Delete multiple notifications
     */
    public void deleteNotifications(User currentUser, List<Long> notificationIds) {
        List<Notification> notifications = notificationRepository.findAllById(notificationIds);
        
        // Verify all notifications belong to current user
        for (Notification notification : notifications) {
            if (!notification.getUser().getId().equals(currentUser.getId())) {
                throw new RuntimeException("You don't have permission to delete some notifications");
            }
        }
        
        notificationRepository.deleteAll(notifications);
    }

    /**
     * Get notification by ID (only if it belongs to current user)
     */
    public NotificationResponse getNotificationById(User currentUser, Long notificationId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Notification not found"));
        
        if (!notification.getUser().getId().equals(currentUser.getId())) {
            throw new RuntimeException("You don't have permission to view this notification");
        }
        
        // Mark as read if not already read
        if (!notification.getIsRead()) {
            notification.markAsRead();
            notificationRepository.save(notification);
        }
        
        return NotificationResponse.from(notification);
    }

    /**
     * Create a single notification for a specific user
     */
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public NotificationResponse createNotification(CreateNotificationRequest request) {
        User targetUser = userRepository.findById(request.getUserId())
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        Notification notification = new Notification(
                request.getTitle(),
                request.getMessage(),
                request.getType(),
                request.getCategory(),
                targetUser
        );
        
        notification.setRelatedEntityId(request.getRelatedEntityId());
        notification.setRelatedEntityType(request.getRelatedEntityType());
        notification.setActionUrl(request.getActionUrl());
        
        notification = notificationRepository.save(notification);
        
        // TODO: Send real-time notification via WebSocket
        
        return NotificationResponse.from(notification);
    }

    /**
     * Send notification to multiple users
     */
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public void createNotificationsForUsers(CreateNotificationRequest request) {
        List<User> targetUsers = userRepository.findAllById(request.getUserIds());
        
        List<Notification> notifications = new ArrayList<>();
        for (User user : targetUsers) {
            Notification notification = new Notification(
                    request.getTitle(),
                    request.getMessage(),
                    request.getType(),
                    request.getCategory(),
                    user
            );
            
            notification.setRelatedEntityId(request.getRelatedEntityId());
            notification.setRelatedEntityType(request.getRelatedEntityType());
            notification.setActionUrl(request.getActionUrl());
            
            notifications.add(notification);
        }
        
        notificationRepository.saveAll(notifications);
        
        // TODO: Send real-time notifications via WebSocket
    }

    /**
     * Send notification to all students in a course
     */
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public void sendNotificationToCourse(Long courseId, CreateNotificationRequest request) {
        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Course not found"));
        
        List<User> students = course.getEnrollments().stream()
                .filter(enrollment -> enrollment.getIsActive())
                .map(enrollment -> enrollment.getStudent())
                .collect(Collectors.toList());
        
        List<Notification> notifications = new ArrayList<>();
        for (User student : students) {
            Notification notification = new Notification(
                    request.getTitle(),
                    request.getMessage(),
                    request.getType(),
                    request.getCategory(),
                    student
            );
            
            notification.setRelatedEntityId(courseId);
            notification.setRelatedEntityType("COURSE");
            notification.setActionUrl(request.getActionUrl());
            
            notifications.add(notification);
        }
        
        notificationRepository.saveAll(notifications);
        
        // TODO: Send real-time notifications via WebSocket
    }

    /**
     * Send notification to all users with a specific role
     */
    @PreAuthorize("hasRole('ADMIN')")
    public void sendNotificationByRole(String roleType, CreateNotificationRequest request) {
        Role role = Role.valueOf(roleType);
        List<User> users = userRepository.findByRole(role);
        
        List<Notification> notifications = new ArrayList<>();
        for (User user : users) {
            Notification notification = new Notification(
                    request.getTitle(),
                    request.getMessage(),
                    request.getType(),
                    request.getCategory(),
                    user
            );
            
            notification.setRelatedEntityId(request.getRelatedEntityId());
            notification.setRelatedEntityType(request.getRelatedEntityType());
            notification.setActionUrl(request.getActionUrl());
            
            notifications.add(notification);
        }
        
        notificationRepository.saveAll(notifications);
        
        // TODO: Send real-time notifications via WebSocket
    }

    /**
     * Create notification for teacher when they create a new assignment
     */
    public void createTeacherAssignmentNotification(User teacher, Long assignmentId, String assignmentTitle, Course course) {
        Notification notification = new Notification(
                "Bài tập đã tạo thành công: " + assignmentTitle,
                "Bạn vừa tạo bài tập mới \"" + assignmentTitle + "\" trong khóa học " + course.getName() + ". Sinh viên sẽ nhận được thông báo.",
                NotificationType.SUCCESS,
                NotificationCategory.ASSIGNMENT,
                teacher
        );
        
        notification.setRelatedEntityId(assignmentId);
        notification.setRelatedEntityType("ASSIGNMENT");
        notification.setActionUrl("/teacher/assignment/" + assignmentId);
        
        notificationRepository.save(notification);
        
        // TODO: Send real-time notifications via WebSocket
    }
    
    /**
     * Create notification for students when new assignment is created
     */
    public void createStudentAssignmentNotification(Long assignmentId, String assignmentTitle, Course course) {
        List<User> students = course.getEnrollments().stream()
                .filter(enrollment -> enrollment.getIsActive())
                .map(enrollment -> enrollment.getStudent())
                .filter(student -> student.getRole() == Role.STUDENT) // Only include actual students, not teachers
                .collect(Collectors.toList());
        
        List<Notification> notifications = new ArrayList<>();
        for (User student : students) {
            Notification notification = new Notification(
                    "Bài tập mới: " + assignmentTitle,
                    "Giảng viên vừa tạo bài tập mới trong khóa học " + course.getName() + ". Hãy kiểm tra và hoàn thành đúng thời hạn.",
                    NotificationType.INFO,
                    NotificationCategory.ASSIGNMENT,
                    student
            );
            
            notification.setRelatedEntityId(assignmentId);
            notification.setRelatedEntityType("ASSIGNMENT");
            notification.setActionUrl("/student/course/" + course.getId() + "/assignment/" + assignmentId);
            
            notifications.add(notification);
        }
        
        notificationRepository.saveAll(notifications);
        
        // TODO: Send real-time notifications via WebSocket
    }

    /**
     * Create notification when assignment is graded
     */
    public void createGradingNotification(User student, Long assignmentId, String assignmentTitle, Double score) {
        Notification notification = new Notification(
                "Bài tập đã được chấm điểm: " + assignmentTitle,
                String.format("Bài tập của bạn đã được chấm điểm. Điểm số: %.2f", score),
                NotificationType.SUCCESS,
                NotificationCategory.GRADING,
                student
        );
        
        notification.setRelatedEntityId(assignmentId);
        notification.setRelatedEntityType("ASSIGNMENT");
        notification.setActionUrl("/student/course/" + getCourseIdFromAssignmentId(assignmentId) + "/assignment/" + assignmentId);
        
        notificationRepository.save(notification);
        
        // TODO: Send real-time notification via WebSocket
    }

    /**
     * Clean up old notifications (can be scheduled)
     */
    @PreAuthorize("hasRole('ADMIN')")
    public void cleanupOldNotifications(int daysToKeep) {
        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(daysToKeep);
        notificationRepository.deleteOldNotifications(cutoffDate);
    }

    /**
     * Send notification about new assignment to all enrolled students
     */
    public void notifyNewAssignment(Assignment assignment) {
        try {
            Course course = assignment.getCourse();
            
            // Get all enrolled students using EnrollmentRepository
            List<Enrollment> enrollments = enrollmentRepository.findByCourseAndIsActiveTrue(course);
            
            List<Notification> notifications = new ArrayList<>();
            for (Enrollment enrollment : enrollments) {
                User student = enrollment.getStudent();
                Notification notification = new Notification(
                        "Bài tập mới: " + assignment.getTitle(),
                        "Bạn có bài tập mới trong khóa học " + course.getName() + 
                        ". Hạn nộp: " + assignment.getEndTime().toString(),
                        NotificationType.INFO,
                        NotificationCategory.ASSIGNMENT,
                        student
                );
                
                notification.setRelatedEntityId(assignment.getId());
                notification.setRelatedEntityType("Assignment");
                notification.setActionUrl("/student/assignments/" + assignment.getId());
                
                notifications.add(notification);
            }
            
            notificationRepository.saveAll(notifications);
            
            // TODO: Send real-time notifications via WebSocket
        } catch (Exception e) {
            // Log the error but don't throw to avoid breaking assignment creation
            System.err.println("Failed to send assignment notifications: " + e.getMessage());
        }
    }

    /**
     * Helper method to get courseId from assignmentId
     */
    private Long getCourseIdFromAssignmentId(Long assignmentId) {
        Assignment assignment = assignmentRepository.findById(assignmentId)
                .orElseThrow(() -> new RuntimeException("Assignment not found"));
        return assignment.getCourse().getId();
    }
}